Enables more tests in CompactionExecutorIT

ddanielr opened the pull request:
Gets all tests working in CompactionExecutorIT. 
Adds additional method to stop compactors by resource group names.

Compaction groups are now created on an as needed basis for each test and cleaned up afterwards without affecting compactions for the root or metadata tables.

ddanielr commented:
The 9th test is `testTooManyDeletes` which seems to be failing because the compaction job is never generated by the 
selector `TooManyDeletesSelector` with the `DeletesSummarizer` summarizer.

I've pushed into the manager code a bit and I'm not sure if the summarizer code is in a working state.

The `stopCompactionGroup` method should probably be switched into something like `stop(ServerType server, String resourceGroup)` but we already have a method signature for a hostname string that would potentially conflict. 

`public synchronized void stop(ServerType server, String hostname) throws IOException {`

keith-turner commented:
> The 9th test is testTooManyDeletes which seems to be failing because the compaction job is never generated by the
selector TooManyDeletesSelector with the DeletesSummarizer summarizer.

I can take a look at this.

keith-turner commented on minicluster/src/main/java/org/apache/accumulo/miniclusterImpl/MiniAccumuloClusterControl.java:
```diff
@@ -254,6 +254,25 @@ public void stop(ServerType server) throws IOException {
+        try {
+          cluster.stopProcessWithTimeout(process, 30, TimeUnit.SECONDS);
+        } catch (ExecutionException | TimeoutException e) {
+          log.warn("TabletServer did not fully stop after 30 seconds", e);
```
```suggestion
          log.warn("Compactor did not fully stop after 30 seconds", e);
```

keith-turner commented:
@ddanielr I opened #4081 about removing selector compactions.  The reason the test if failing is that selector compactions are not implemented in elasticity.  Looking into this I am thinking we should remove the functionality rather than implement it.

I modified the test not to wait on the selector compaction and it works.  The user compactions that select files work.  If we know we are going to take the path outlined in #4081 the we could update the test to remove setting the selector plugin on the table.

```diff
diff --git a/test/src/main/java/org/apache/accumulo/test/compaction/CompactionExecutorIT.java b/test/src/main/java/org/apache/accumulo/test/compaction/CompactionExecutorIT.java
index a7309cb693..92c0c0f49f 100644
--- a/test/src/main/java/org/apache/accumulo/test/compaction/CompactionExecutorIT.java
+++ b/test/src/main/java/org/apache/accumulo/test/compaction/CompactionExecutorIT.java
@@ -358,9 +358,9 @@ public class CompactionExecutorIT extends SharedMiniClusterBase {
       assertEquals(2, getFiles(client, "tmd_control2").size());
       assertEquals(2, getFiles(client, "tmd_control3").size());
 
-      while (getFiles(client, "tmd_selector").size() != 0) {
-        Thread.sleep(100);
-      }
+      //while (getFiles(client, "tmd_selector").size() != 0) {
+      //  Thread.sleep(100);
+      //}
 
       assertEquals(2, getFiles(client, "tmd_control1").size());
       assertEquals(2, getFiles(client, "tmd_control2").size());
```

ddanielr commented:
> @ddanielr I opened #4081 about removing selector compactions. The reason the test is failing is that selector compactions are not implemented in elasticity. Looking into this I am thinking we should remove the functionality rather than implement it.
> 
> I modified the test not to wait on the selector compaction and it works. The user compactions that select files work. If we know we are going to take the path outlined in #4081 the we could update the test to remove setting the selector plugin on the table.
> 
> ```diff
> diff --git a/test/src/main/java/org/apache/accumulo/test/compaction/CompactionExecutorIT.java b/test/src/main/java/org/apache/accumulo/test/compaction/CompactionExecutorIT.java
> index a7309cb693..92c0c0f49f 100644
> --- a/test/src/main/java/org/apache/accumulo/test/compaction/CompactionExecutorIT.java
> +++ b/test/src/main/java/org/apache/accumulo/test/compaction/CompactionExecutorIT.java
> @@ -358,9 +358,9 @@ public class CompactionExecutorIT extends SharedMiniClusterBase {
>        assertEquals(2, getFiles(client, "tmd_control2").size());
>        assertEquals(2, getFiles(client, "tmd_control3").size());
>  
> -      while (getFiles(client, "tmd_selector").size() != 0) {
> -        Thread.sleep(100);
> -      }
> +      //while (getFiles(client, "tmd_selector").size() != 0) {
> +      //  Thread.sleep(100);
> +      //}
>  
>        assertEquals(2, getFiles(client, "tmd_control1").size());
>        assertEquals(2, getFiles(client, "tmd_control2").size());
> ```

I removed the code for the selector compactions and now all tests are working. 
I think #4081 is the correct way to go since it's moving all of those custom compaction functions under user control.

If we come to a different decision with #4081 then we could revisit adding the logic back to this test.

keith-turner commented on minicluster/src/main/java/org/apache/accumulo/miniclusterImpl/MiniAccumuloClusterControl.java:
```diff
@@ -254,6 +254,25 @@ public void stop(ServerType server) throws IOException {
+        try {
+          cluster.stopProcessWithTimeout(process, 30, TimeUnit.SECONDS);
+        } catch (ExecutionException | TimeoutException e) {
+          log.warn("Compactor did not fully stop after 30 seconds", e);
```
This is following the pattern of other code, so I think this is fine here.  This general pattern seems like it could cause trouble.  If a test wants to stop something and runs into this and fails later because it was not stopped could make the test harder to debug.

ddanielr replied:
> This is following the pattern of other code, so I think this is fine here. This general pattern seems like it could cause trouble. If a test wants to stop something and runs into this and fails later because it was not stopped could make the test harder to debug.

I agree that it wouldn't be immediately obvious that the shutdown failing caused an error. 

To help with that, we could either throw an exception on failed process stop or this method could return an exit code. 

I'm leaning towards the exit code route since it cuts down on the amount of try/catch statements in a given test and also allows the test to ignore the output of this method if they wanted.

Do you have an opinion?

keith-turner replied:
If there are frequent cases of things wanting to ignore a failure to stop then a returning an exit code makes sense.  If not then throwing a runtime exception would cause things to fail and not introduce new try/catch blocks in the code.  Would not want to throw a checked exception, that seems like it would be annoying for this case w/o any benefit.

keith-turner commented on test/src/main/java/org/apache/accumulo/test/compaction/CompactionExecutorIT.java:
```diff
@@ -67,36 +69,49 @@
   public static class TestPlanner implements CompactionPlanner {
 
+    private static class ExecutorConfig {
+      String name;
+      String type;
```
Does not seem like these two fields are read, but not sure.  If they are not read could they be dropped?

ddanielr replied:
Yes they can be dropped.

keith-turner commented on test/src/main/java/org/apache/accumulo/test/compaction/CompactionExecutorIT.java:
```diff
@@ -180,6 +212,18 @@ public static void teardown() {
+      try {
+        getCluster().getClusterControl().stopCompactorGroup(group);
+      } catch (Exception e) {
+        // Exception is encountered but ignored.
```
Why ignore this?  Was wondering if it would be useful to log it.

ddanielr replied:
Added logging statement for the exception.

keith-turner commented on test/src/main/java/org/apache/accumulo/test/compaction/CompactionExecutorIT.java:
```diff
@@ -180,6 +212,18 @@ public static void teardown() {
+    var iter = compactionGroups.iterator();
+    while (iter.hasNext()) {
+      var group = iter.next();
+      getCluster().getConfig().getClusterServerConfiguration().addCompactorResourceGroup(group, 0);
```
If its quick to explain, I am curious about why this is called with zero. If its not quick to explain let me know and I can look at the code.

ddanielr replied:
We set the value to zero so the next call to `getCluster().getClusterControl().start(ServerType.COMPACTOR);` doesn't start up old compactors.

ddanielr replied:
The other reason why we might have a group set to zero is because the miniAccumuloClusterImpl goes through on `start()` and creates compactorResourceGroups with a single compactor for each "group" it can find in the "executors" config.

However, there might be cases where we don't want that compactor to run by default, so we override that with a `0` setting. 

https://github.com/apache/accumulo/blob/feadf10305fec490a32211606d87a06a5b398d65/minicluster/src/main/java/org/apache/accumulo/miniclusterImpl/MiniAccumuloClusterImpl.java#L638-L642

keith-turner reviewed (approved):
I think this looks good, could maybe simplify the test json as a I made a comment about.

Did you consider moving to having each test method run its own mini accumulo instead of stoppping compactors between test?

ddanielr commented:
> I think this looks good, could maybe simplify the test json as a I made a comment about.

I simplified the json and added a runtime exception for the stopCompactorGroup test.
 
> Did you consider moving to having each test method run its own mini accumulo instead of stoppping compactors between test?

I did but I decided against that because spinning compaction resources up and down against a long-lived cluster is closer to a "real-life" test scenario then testing the compaction services in isolation. 

Also, an extra step is required for tests that change the compaction resources mid-test for the AccumuloClusterHarness.

You must cast the config in order to access the groups directly and change their values.

```
MiniAccumuloClusterImpl mini = (MiniAccumuloClusterImpl) getCluster();
mini.getConfig().getClusterServerConfiguration().addCompactorResourceGroup("group1", 4);
mini.getClusterControl().start(ServerType.COMPACTOR);
```
